### ISSUE TITLE
Slower inference time running MobileNet V3  when compared to PyTorch Mobile

### ISSUE THREAD

**Issue by laoluani** [pytorch/executorch#4005]
I'm currently accessing the performance of Exectorch compared to PyTorch Mobile. When running the MobileNet V3 Small model from the Pytorch vision model hub. I find that it's slower than running on PyTorch Mobile v1.12.1 on Android the Samsung S20 Ultra.
I have taken the android Executorch Demo app and integerated PyTorch Mobile and run the model side by side.

On average running on Executorch I get a forward pass time of around `30ms` and `8ms` on PyTorch. 

**PyTorch Mobile set up**
Here's the script used to do the conversion: 
``` 
from torchvision import models
import torch
from torch.utils.mobile_optimizer import optimize_for_mobile

mv3_small = models.mobilenet_v3_small(weights=models.MobileNet_V3_Small_Weights.IMAGENET1K_V1)
input = torch.randn((1, 3, 224, 224))
mv3_small.eval()
torchscript_model = torch.jit.trace(model, input)
torchscript_model._save_for_lite_interpreter("mv3.ptl")
```
I'm not applying any extra optimisations in the mobile model conversion

**Executorch set up** 
This follows the XNNPACK demo branched from v.0.2.0, but altered to use the MobileNet V3 small. 
Here's output from XNNPACK lowering process 
```
graph():
    %arg210_1 : [num_users=1] = placeholder[target=arg210_1]
    %lowered_module_0 : [num_users=1] = get_attr[target=lowered_module_0]
    %executorch_call_delegate : [num_users=1] = call_function[target=torch.ops.higher_order.executorch_call_delegate](args = (%lowered_module_0, %arg210_1), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%executorch_call_delegate, 0), kwargs = {})
    %aten_view_copy_default : [num_users=1] = call_function[target=executorch.exir.dialects.edge._ops.aten.view_copy.default](args = (%getitem, [1, 576]), kwargs = {})
    %lowered_module_1 : [num_users=1] = get_attr[target=lowered_module_1]
    %executorch_call_delegate_1 : [num_users=1] = call_function[target=torch.ops.higher_order.executorch_call_delegate](args = (%lowered_module_1, %aten_view_copy_default), kwargs = {})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%executorch_call_delegate_1, 0), kwargs = {})
    return (getitem_1,)
```

I assume In these circumstances both Pytorch Mobile and Executorch are both using the XNNPACK backed where possible. So I would expect forwards pass times to be similar. Executorch has the added advantage of the  applying the extra preprocessing to the model when lowering.

I'm guessing the general optimisations that `torch.jit.trace` applies to model is out performing the specific optimisations of the XNNPACK preprocess step.

Any recommendations for speed ups?

**Comment by guangy10** [comment]
@laoluani Thanks you for sharing the pref experiment with us. 

> both Pytorch Mobile and Executorch are both using the XNNPACK backed where possible

I don't even see you call optimize_for_mobile for the PyTorch Mobile model, so it seems just scripts the model. @kimishpatel or @cccclai can clarify how the old torchscript based solution works.

**Comment by cbilgin** [comment]
> I'm guessing the general optimisations that torch.jit.trace applies to model is out performing the specific optimisations of the XNNPACK preprocess step.

You should actually be seeing speedups with ExecuTorch for MV3 compared to torchscript. Can you share how you lowered to xnnpack as well?

**Comment by laoluani** [comment]
I realised that I hadn't turned on the release flag for my ExecuTorch  builds, once enabled I got results much closer to PyTorch Mobile.

I Rana  test over 500 iterations with new random inputs on every pass
```
pytorch mobile avg: 26.05, max: 57, min: 8, std deviation: 11.24
executorch avg: 26.32, max: 75, min: 5, std deviation: 10.22
```

Here's how I lowered the model: 

```
import torch
import torchvision.models as models
import logging

from torch.export import export, ExportedProgram
from torchvision.models.mobilenetv2 import MobileNet_V2_Weights
from executorch.backends.xnnpack.partition.xnnpack_partitioner import XnnpackPartitioner
from executorch.exir import EdgeProgramManager, ExecutorchProgramManager, to_edge
from executorch.exir.backend.backend_api import to_backend

mobilenet_v3 = models.mobilenet_v3_small(weights=models.MobileNet_V3_Small_Weights.IMAGENET1K_V1).eval()
sample_inputs = (torch.randn(1, 3, 224, 224), )

exported_program: ExportedProgram = export(mobilenet_v3, sample_inputs)
edge: EdgeProgramManager = to_edge(exported_program)

logging.info(f"Exported graph:\n{edge.exported_program()}")

# Lower the model to XNNPACK backend
edge = edge.to_backend(XnnpackPartitioner())
logging.info(f"Lowered graph:\n{edge.exported_program()}")

exec_prog = edge.to_executorch()

with open("mv3-xnnpack.pte", "wb") as file:
    exec_prog.write_to_file(file) 
```

 Here are the build commands:

```
cmake . -DCMAKE_INSTALL_PREFIX=/build \
    -DCMAKE_TOOLCHAIN_FILE="$ANDROID_NDK/build/cmake/android.toolchain.cmake" \
    -DANDROID_ABI="$ANDROID_ABI" \
    -DEXECUTORCH_BUILD_XNNPACK=ON \
    -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON \
    -DEXECUTORCH_BUILD_EXTENSION_DATA_LOADER=ON \
    -DEXECUTORCH_BUILD_OPTIMIZED=ON \
    -DCMAKE_BUILD_TYPE=Release \
    -B/build

find "/build" -type f -exec sed -i 's/-fno-exceptions/-fexceptions/g' {} +

cmake --build /build -j16 --target install --config Release

cmake extension/android \
  -DCMAKE_TOOLCHAIN_FILE="${ANDROID_NDK}"/build/cmake/android.toolchain.cmake \
  -DANDROID_ABI="${ANDROID_ABI}" \
  -DCMAKE_INSTALL_PREFIX=/build \
  -DCMAKE_BUILD_TYPE=Release \
  -B/build/extension/android

cmake --build /build/extension/android -j16 --config Release
```

Are these the expected results and are there any improvements that can be made?

**Comment by GregoryComer** [comment]
Thanks for the info. I will try to reproduce the numbers and get back to you.

**Comment by kirklandsign** [comment]
Hi @laoluani have you tried running the same model with cmake-out-android/backends/xnnpack/xnn_executor_runner? Does it produce the same inference time as your integrated app?

**Comment by kirklandsign** [comment]
Hi @laoluani you probably need to use the XNNPACK runner.

**Comment by laoluani** [comment]
Yes that worked for me! What's the best way to get timing info out? 

Looks like I'm just getting raw scores.

```
z3q:/data/local/tmp $ /data/local/tmp/xnn_executor_runner --model_path=./mv3-xnnpack.pte                                                                                 
Output 0: tensor(sizes=[1, 1000], [
  -0.0297435, -0.116339, 0.234517, -0.115163, 0.28406, 1.33266, -1.2022, -0.418206, -0.861483, 0.962639, 
  2.05277, 0.0322932, -0.672343, -0.137665, -0.785477, -0.191068, 0.478215, -1.26891, -0.0889226, -0.0928001, 
  -1.35861, 3.53779, 1.71552, 1.3614, -0.172415, 0.221101, 1.33974, 0.580504, 1.56505, -0.204694, 
  -1.91138, 0.0594309, -0.480634, -0.466428, 0.812006, -1.7187, 0.209461, -1.64942, 2.64706, -1.81845, 
  0.0644182, 0.913044, 1.08294, 2.03153, 0.366071, 1.04362, -0.548808, 2.08426, -1.41956, -1.71163, 
  -1.92737, 0.822696, 1.65539, 0.988034, 0.465189, -1.36933, -0.598565, -0.559072, -0.845721, 1.56357, 
  0.656557, -0.246826, 0.556268, 1.71325, 0.38631, 0.838485, 1.38649, -0.594456, 1.44696, -0.336695, 
  0.817791, 2.16338, 0.452428, 2.40616, -0.203028, 1.58678, -0.00467801, 0.391145, 3.28668, 1.32205, 
  2.26962, 0.525644, 0.393954, 2.02201, -1.76368, -0.793006, 0.871535, 0.615611, -0.240767, 0.662297, 
  -0.822543, 0.770243, 1.65924, -0.345425, -0.102899, 0.538089, 0.0803897, -1.50696, 0.749941, 0.125525, 
  ...,
  0.978575, 2.59865, 3.61896, 0.340622, 0.44494, 0.760727, 2.25554, 0.673822, 0.773279, 0.901685, 
  1.19766, 0.744912, 1.02076, 0.317082, 1.93721, -0.230186, 2.05909, 1.14174, -0.71805, 2.01184, 
  1.97479, 2.68669, 0.0470679, 0.85921, -1.01457, 1.66463, 1.62474, 0.308362, -0.830093, 0.904799, 
  2.00759, -0.694682, -0.0069273, -0.317874, 0.0855309, 0.020355, -0.0538518, -0.501041, 0.918039, -1.4842, 
  1.68228, -0.304506, 0.818993, -0.442584, -1.71591, 0.721102, -1.11913, -0.560189, 0.792763, 1.72536, 
  0.0979954, 0.635417, 0.140096, -0.858204, 0.531698, -1.45714, 0.525842, -0.311962, -0.148772, 0.240123, 
  0.233015, 0.418201, 0.319395, -1.34773, 0.813186, -0.508101, 1.48234, 0.756897, 0.994117, 2.93617, 
  0.502231, 0.777379, 0.751076, -1.91975, 1.34087, 0.581285, 1.6911, 1.16569, 2.25252, 0.19116, 
  0.948301, 0.508456, 0.229856, -0.155784, 0.170956, 0.545924, -0.805998, 0.195962, -0.0608273, -0.785059, 
  -0.438103, -1.82035, -0.877346, -0.129996, -0.643686, 0.254722, -0.689362, -0.311216, 0.469063, 1.01164, 
])```

**Comment by kirklandsign** [comment]
Does adb log have timestamp?

**Comment by laoluani** [comment]
So using the linux `time` command I got the following:  `0m00.15s real     0m00.45s user     0m00.04s system`. So looks like the total time was about 15ms. Is the executor runner doing a single forward or multiple?

If this is a single pass, this falls within the range of values I got whilst testing. What's the difference between the executor runner and the actual ExecuTorch runtime?

**Comment by kirklandsign** [comment]
> What's the difference between the executor runner and the actual ExecuTorch runtime?

There isn't. executor runner is a binary entry point for the runtime library

**Comment by mcr229** [comment]
hi @laoluani  it seems like the timing issue was resolved here? If not there is further tuning you can use such as the thread count flag in the xnn_executor_runner. 

If the performance issues still persist, feel free to reopen this task.
